Automated subscription simulcast switching based on remb

JanFellner opened the pull request:
This PR is a followup from the original one of @sjkummer (https://github.com/meetecho/janus-gateway/pull/2735)

Fine tuned the switching and its working smothly but currenty lacks in probing. Therefore as draft PR to discuss where and how to implement probing.

The context and whats currently implemented.
- Via the api i can provide a simulcast_bitrates array to janus for the publisher. Janus then know the bitrates the different layers of the publisher will consume.
- When creating a subscription on the publisher those bitrate values are currently copied to the subscription (to not need to mutex lock the publisher for every remb event)
- On an inbound remb value we determine if the value is fitting the current layer, is below, or above the next higher layer.
- In case the bitrate continuously does not fit what we need the layer is reduced (the logic does not change what the subscription shall transport, the value is an overlay to what has been requested through the API)
- In case the bitrate is higher than the next layer this overlay reduction value is increased and therefore a switching to a higher layer is possible when the client did request a higher one before.

Whats open
- The logic works smoothly if i come from high and the bitrate drops.
- It does not work if i come from a lower bitrate and the REMB value increases.

The reason for that is that the REMB does not increase to a value that covers the next layer.
if i have layers configured like this: 150/500/1200kbit
REMB does no got above 500 if i am on the lowest layer (max is about 380kbit).
The same happens for the middle layer, REMB does not go above 1200kbit (max is about 950kbit).

Therefore probing would be needed.
The question is therefore:
- Where would be the best place to implement probing.

JanFellner committed:
CC-2137 Taken over changes from sjkummer

JanFellner committed:
CC-2137 Improvements, do not change what the client requested, the remb is just a limit on what the client wants to have

JanFellner committed:
CC-2137 logic complete, probing is missing

JanFellner committed:
CC-2137 Also handle temporal layers in ramping up ramping down

fippo commented on plugins/janus_videoroom.c:
```diff
@@ -5501,6 +5504,97 @@ void janus_videoroom_incoming_rtp(janus_plugin_session *handle, janus_plugin_rtp
 
+/* Retrieve the current position in the ramp up ramp down process
+ * Ramp positions:
+ * Spatial 0 (L) - Temporal 15fps -> 0
```
i think you'll run into issues with this as only chrome restricts itself to two temporal layers while other browsers will use three (which remains the default of libwebrtc)

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

JanFellner committed:
CC-2137 finalized the first idea

JanFellner committed:
CC-2137 Added that we need VP9 and put the missing remb in the general statistics to the top

JanFellner committed:
CC-2137 refactored code to have all the remb stream switcher values in a dedicated sub struct of the simulcast data (just to have it properly separated)

JanFellner committed:
CC-2137 Moved calculation of the next higher or lower ramp position into one function.

JanFellner committed:
CC-2137 Detection of failed ramp ups

JanFellner committed:
CC-2137 just improve/changed readability of the develop logoutput

JanFellner committed:
CC-2137 fixed finding by @roisasson

JanFellner committed:
CC-2137 as discussed and suggested by @lminiero put the whole implementation into a dedicated file to have it loosen from the video room and reusable for the other plugins.

JanFellner committed:
CC-2137 changed some variable names to better match the content

JanFellner committed:
CC-2137 reset the internal state if the client is requesting a lower ramp position than the one we currently send

JanFellner committed:
CC-2137 Several fixes as i mixed up current vs requested ramp position in the last commits.
Have an own flag that tells me wether the last_bitrate value is valid or not (did we already receive a valid remb package or not)

JanFellner committed:
CC-2137 Improve diagnostic log to see the remb bitrate values plugin handle related
- add an option to switch on/off debugging
- added tests that have been done (which browser version and whats the finding)

JanFellner committed:
CC-2137 add api properties to configure how the subscription shall adopt againt the remb.

JanFellner committed:
CC-2137 Via Subscribe, Configure, Swtich it is now possible to advice janus how to handle the remb adoption from the subscriber side. This is either ignore, ramp_up, or start_high.
ignore - no remb adoption
ramp_up - the adoptoin starts on the lowest simulcast layer with temporal 15fps and ramps up (for ff)
start_high - it starts with what the subscriber wants to have (as chrome and safari delivers a pretty good remb value right from the start)

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

JanFellner committed:
CC-2137 Adopted coding style

JanFellner committed:
CC-2137 fixed a missing coding style recommondation

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

lminiero committed:
Grow buffer as needed when generating SDPs (fixes #2791, see #2793)

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

JanFellner committed:
Merge branch 'sdp-size' of https://github.com/meetecho/janus-gateway into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb

JanFellner committed:
Merge branch 'meetecho:master' into feature/CC-2137-automated-subscription-simulcast-switching-based-on-remb
